syntax = "proto3";

import "corpc_option.proto";
import "common.proto";
import "inner_common.proto";

option cc_generic_services = true;

package wukong.pb;

message LoadRoleDataRequest {
    uint32 serverId = 1; // 物理服编号
    uint32 lToken = 2;
    uint32 roleId = 3;
}

message LoadRoleDataResponse {
    uint32 errCode = 1;
    uint32 serverId = 2; // 游戏内角色所属逻辑服号
    bytes data = 3;
}

message SyncData {
    string key = 1;
    bytes value = 2;
}

message SyncRequest {
    uint32 serverId = 1; // 物理服编号
    uint32 lToken = 2;
    uint32 roleId = 3;
    repeated SyncData datas = 4;    // 增加或修改的数据
    repeated string removes = 5;    // 删除的数据
}

message RSHeartbeatRequest {
    uint32 serverId = 1; // 物理服编号
    uint32 roleId = 2;
    uint32 lToken = 3;
}

// 玩家游戏对象所在服务器需要实现GameService服务
service RecordService {
    option (corpc.global_service_id) = 4;

    // 关闭record服
    rpc shutdown(corpc.Void) returns(corpc.Void)
    {
        option (corpc.not_care_response) = true; // default is false
    }

    // 获取在线玩家数，用于负载均衡（超时时间1秒，超时表示负载过高，后续不被继续分配）
    rpc getOnlineCount(corpc.Void) returns(OnlineCounts)
    {
        option (corpc.need_coroutine) = true; // default is false
        option (corpc.timeout) = 1000; // default is 0
    }

    // 加载指定角色
    rpc loadRoleData(LoadRoleDataRequest) returns(LoadRoleDataResponse)
    {
        option (corpc.need_coroutine) = true; // default is false
    }

    // 保存数据
    rpc sync(SyncRequest) returns(BoolValue)
    {
        option (corpc.need_coroutine) = true; // default is false
        option (corpc.timeout) = 2000; // default is 0
    }

    // 玩家游戏对象向路由对象定时心跳查看路由对象是否存在（超时时间2秒，负载过高心跳失败会导致游戏对象销毁）
    rpc heartbeat(RSHeartbeatRequest) returns(BoolValue)
    {
        option (corpc.need_coroutine) = true; // default is false
        option (corpc.timeout) = 2000; // default is 0
    }
}

service InnerRecordService {
    option (corpc.global_service_id) = 4;

    // 关闭record服
    rpc shutdown(corpc.Void) returns(corpc.Void)
    {
        option (corpc.not_care_response) = true; // default is false
    }

    // 获取在线玩家数，用于负载均衡（超时表示负载过高，后续不被继续分配）
    rpc getOnlineCount(corpc.Void) returns(Uint32Value)
    {
        option (corpc.timeout) = 500; // default is 0
    }

    // 加载指定角色
    rpc loadRoleData(LoadRoleDataRequest) returns(LoadRoleDataResponse)
    {
        option (corpc.need_coroutine) = true; // default is false
    }

    // 保存数据
    rpc sync(SyncRequest) returns(BoolValue)
    {
        option (corpc.timeout) = 2000; // default is 0
    }

    // 玩家游戏对象向路由对象定时心跳查看路由对象是否存在（超时时间2秒，负载过高心跳失败导致游戏对象销毁）
    rpc heartbeat(RSHeartbeatRequest) returns(BoolValue)
    {
        option (corpc.timeout) = 2000; // default is 0
    }
}